projects/output-and-error-handling/README.md

```diff
@@ -0,0 +1,195 @@
+
+```go
+func readPassword() (string, error) {
+    password, err := os.ReadFile(".some-file")
+    if err != nil {
+        return "", fmt.Errorf("failed to read password file: %w", err)
+    }
+    return string(password), nil
+}
+```
+
+You can learn more about creating and wrapping errors in Go by reading [this article](https://earthly.dev/blog/golang-errors/).
+
+#### Working around the error to recover from it
+
+Sometimes, an error may be expected, or may be recoverable. For instance, suppose we have some expensive computation we want to do, but which may have already been done and saved to a file. We may try to read the file, but if we encounter an error that it doesn't exist, we may know how to compute the answer we need instead.
+
+This kind of behaviour will depend entirely on the problem domain you're solving, and there isn't really a general rule for when it's appropriate.
+
+#### Terminating the program completely
+
+A lot of the time, when we run into errors, there's nothing we can do about them.
+
+If we're running a server, and the error happened when processing one request, normally we don't want to terminate our program - we just want to respond to that request saying an error happened, but keep trying to process other requests.
+
+Other times, for instance when first starting up a server, or when writing a program that isn't a server but just does a one-off task, it may make sense to terminate our program, and exit (with a non-`0` status code).
+
+_Where_ we do this, however, is worth thinking about. We generally don't want to call `os.Exit` from anywhere _except_ our `main` function.
+
+There are a few reasons for this:
+1. If we call `os.Exit`, there's no way any code can handle that or recover. Let's say we started calling `os.Exit` in some other function - it's possible we'll end up in the future calling that function from a request handler, and we'll end up terminating the whole server just because one request couldn't be handled. This will probably cause an outage, because no one will be able to talk to our server any more.
+1. When writing unit tests, we generally don't want our program to exit. But if you call `os.Exit` inside a unit test, it will stop running. In general, we never want to call `os.Exit` from any code which is called from a test.
+
+While you may write an `os.Exit` call in some other function, thinking it's only ever called from places it's ok to call `os.Exit`, code changes a lot over time, and you may find yourself or someone else adding calls to functions from other places without realising that your function isn't safe to be called everywhere. The easiest way to avoid this is to use the rule: only ever call `os.Exit` from your `main` function - everything else should propagate any errors they encounter.
+
+#### Ignoring the error
+
+Sometimes an error actually doesn't matter at all, and can just be ignored. This is rare, and you should be wary if you think this is the case.
```

Comment:
it could be nice to link to this paper: https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-yuan.pdf
a great read from a few years back, basically shows that a huge percentage of distirbuted systems bugs are due to ignoring errors, or improper handling of some kind.